### PR TITLE
[LayerNorm] Add fused XPU LayerNorm kernels for the eager path

### PR DESCRIPTION
**Pull request by mganczarenko**
## Summary

Adds native XPU LayerNorm kernels so models using plain `nn.LayerNorm`
(GPT-J, GPT-2, OPT, Nemotron, etc.) get a fused kernel on the
`--enforce-eager` path instead of falling back to torch-xpu-ops' native
`layer_norm`, which measured well below achievable bandwidth on this
hardware. Mirrors the structure of the existing `rms_norm_kernel` /
`fused_add_rms_norm_kernel` in this file — vectorized + scalar-fallback
variants, same dispatch pattern.

LayerNorm needs a second reduction pass that RMSNorm doesn't (mean, then
centered variance), so the kernel carries two `local_accessor<float>`
slots instead of one, and both weight and bias are optional (RMSNorm
only has weight).

Four ops:

- `layer_norm` / `fused_add_layer_norm` — generic, `weight=None`/`bias=None`
  allowed (covers `elementwise_affine=False` callers too).
- `nemotron_layer_norm` / `fused_add_nemotron_layer_norm` — weight
  required, bias optional. Reuses the `weight_bias` trick from
  `gemma_rms_norm` (#532) for Nemotron's `(1 + weight)` offset, folded
  into the kernel instead of recomputed every forward call.

## Performance (direct op call, bf16, B70)

Vectorized path (hidden_size=4096, divisible by the vector width):

| tokens | native `F.layer_norm` | sycl kernel | speedup |
|---|---|---|---|
| 1 | 8.35us | 3.60us | 2.32x |
| 128 | 9.63us | 3.70us | 2.60x |
| 512 | 28.20us | 5.90us | 4.78x |
| 1024 | 51.73us | 10.83us | 4.78x (peak) |
| 2048 | 103.91us | 56.61us | 1.84x |

Scalar-fallback path (hidden_size=4097, not divisible by the vector
width — forces the non-vectorizable branch):

| tokens | native `F.layer_norm` | sycl kernel | speedup |
|---|---|---|---|
| 1 | 11.01us | 5.90us | 1.87x |
| 128 | 11.87us | 6.53us | 1.82x |
| 512 | 34.43us | 16.84us | 2.05x |
| 1024 | 61.78us | 32.21us | 1.92x |
| 2048 | 136.88us | 101.19us | 1.35x |

Wins at every size tested on both paths. (A prior Triton-kernel attempt
at this same gap regressed below ~300 tokens — 0.4-0.97x — due to fixed
Triton launch overhead on XPU; this SYCL version doesn't have that tax
since it dispatches through the same warm `torch.ops` path native ops
use.)

The non-vectorizable branch dispatches to the same kernel template at
`VEC_SIZE=1` rather than a separate hand-written scalar specialization
(`vec_n_t<scalar_t, 1>` degenerates to a single-element load/store) —
saves ~185 lines vs. the `rms_norm_kernel` convention of a fully
duplicated `VEC_SIZE=0` class, verified bit-identical and perf-neutral
against that convention on both tables above.

## Testing

Ported into `tests/test_layernorm.py`'s real `opcheck`-based harness
(`test_layer_norm`, `test_nemotron_layer_norm`), reusing the existing
`HIDDEN_SIZES` list (which already includes several sizes not divisible
by the vector width, so both the vectorized and scalar-fallback paths
get exercised) plus a new `HAS_BIAS` axis. Full parametrized sweep:
**2112 passed, 0 failed.**

### CODE REVIEW

File: csrc/torch_bindings.cpp [hunk 1]
```diff
@@ -48,6 +48,34 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
...
+  // the +1 weight offset into the kernel (same trick as gemma_rms_norm).
+  ops.def(
+      "nemotron_layer_norm(Tensor! result, Tensor! input, Tensor! weight, "
+      "Tensor? bias, float epsilon) -> ()");
+  ops.impl("nemotron_layer_norm", torch::kXPU, &nemotron_layer_norm);
+
+  // In-place fused Add and Nemotron LayerNorm.
+  ops.def(
+      "fused_add_nemotron_layer_norm(Tensor! input, Tensor! residual, "
+      "Tensor! weight, Tensor? bias, float epsilon) -> ()");
```
[the remark this answers is not in the document]
Comment (mganczarenko) [reply csrc/torch_bindings.cpp]: Already commited

File: tests/register_ops.py [hunk 2]
```diff
@@ -38,6 +38,38 @@ def fused_add_gemma_rms_norm(input: torch.Tensor, residual: torch.Tensor,
     torch.ops._C.fused_add_gemma_rms_norm(input, residual, weight, epsilon)
 
 
+def layer_norm(out: torch.Tensor, input: torch.Tensor,
+               weight: Optional[torch.Tensor], bias: Optional[torch.Tensor],
+               epsilon: float) -> None:
+    # Mirrors rms_norm's own .contiguous() call above; the kernel already
+    # handles this internally, kept here for parity.
```
[the remark this answers is not in the document]
Comment (mganczarenko) [reply tests/register_ops.py:45]: Addressed by changing the condition to .is_contiguous()

File: csrc/layernorm.cpp [hunk 3]
```diff
@@ -882,3 +1288,153 @@ void fused_add_gemma_rms_norm(
...
+  }
+  TORCH_CHECK(input.stride(-1) == 1);
+  const bool has_weight = weight.has_value();
+  const bool has_bias = bias.has_value();
+  if (has_weight) {
+    TORCH_CHECK(weight->is_contiguous());
+  }
+  if (has_bias) {
+    TORCH_CHECK(bias->is_contiguous());
+  }
```
[the remark this answers is not in the document]
Comment (mganczarenko) [reply csrc/layernorm.cpp:1327]: Addressed by `717e0ed`

**Comment by jikunshang** [comment]
please fix pre-commit issue. thanks.

**Comment by mganczarenko** [comment]
@jikunshang fixed

**Comment by mganczarenko** [comment]
@jikunshang, @mayuyuace, could you help with review? thanks a lot

**Comment by mayuyuace** [comment]
Can you please add benchmark file also?

**Comment by mganczarenko** [comment]
@mayuyuace Sure, here is the script:

```python
import time

import torch
import torch.nn.functional as F

import vllm_xpu_kernels._C  # noqa: F401

device = "xpu"
torch.manual_seed(0)


def bench(fn, iters=300, warmup=30):
    for _ in range(warmup):
        fn()
    torch.xpu.synchronize()
    start = time.perf_counter()
    for _ in range(iters):
        fn()
    torch.xpu.synchronize()
    return (time.perf_counter() - start) / iters * 1e6


hidden_size = 4096  # use 4097 for the scalar-fallback path (VEC_SIZE=1)
print(f"{'tokens':<10}{'native (us)':<14}{'sycl (us)':<14}{'speedup':<10}")
for num_tokens in [1, 8, 32, 64, 128, 256, 384, 512, 768, 1024, 1536, 2048]:
    x = torch.randn(num_tokens, hidden_size, dtype=torch.bfloat16, device=device)
    weight = torch.randn(hidden_size, dtype=torch.bfloat16, device=device)
    bias = torch.randn(hidden_size, dtype=torch.bfloat16, device=device)
    eps = 1e-5
    out = torch.empty_like(x)

    def native():
        return F.layer_norm(x, (hidden_size,), weight, bias, eps)

    def sycl():
        torch.ops._C.layer_norm(out, x, weight, bias, eps)
        return out

    n = bench(native)
    s = bench(sycl)
    print(f"{num_tokens:<10}{n:<14.2f}{s:<14.2f}{n/s:<10.2f}x")
```

Notes for anyone reproducing:
- Wrote this standalone script during kernel development to check the kernel's raw compute performance in isolation, decoupled from the `CustomOp` dispatch/allocation overhead in `tests/ops/layernorm_op.py` — that machinery wasn't the point of the check, the kernel's own numbers were.
- These are single-run means (300 iters, 30 warmup), no repeated-run median/spread — the `tokens=1` row is noisiest since actual compute time there is tiny enough that host-scheduling jitter dominates. Reran it independently and got the same numbers within ~1% at 128+ tokens; `tokens=1` varied more run to run.
- This measures the raw op call — no `CustomOp` dispatch overhead, and the output buffer is preallocated once and reused across iterations. If you benchmark through the model-integration path (`XPULayerNorm` / `nn.LayerNorm` swap), expect a smaller speedup — the dispatch layer and a per-call output allocation add fixed overhead that's most visible at small `hidden_size`/token counts.
- `torch.nn.functional.layer_norm` is the correct native comparison point here — it's the real fallback this kernel replaces, not the test suite's unfused fp32 reference implementation.

**Review by jikunshang — APPROVED** [review]

**Comment by mayuyuace** [comment]
> @mayuyuace Sure, here is the script:
> 
> ```python
> import time
> 
> import torch
> import torch.nn.functional as F
> 
> import vllm_xpu_kernels._C  # noqa: F401
> 
> device = "xpu"
> torch.manual_seed(0)
> 
> 
> def bench(fn, iters=300, warmup=30):
>     for _ in range(warmup):
>         fn()
>     torch.xpu.synchronize()
>     start = time.perf_counter()
>     for _ in range(iters):
>         fn()
>     torch.xpu.synchronize()
>     return (time.perf_counter() - start) / iters * 1e6
> 
> 
> hidden_size = 4096  # use 4097 for the scalar-fallback path (VEC_SIZE=1)
> print(f"{'tokens':<10}{'native (us)':<14}{'sycl (us)':<14}{'speedup':<10}")
> for num_tokens in [1, 8, 32, 64, 128, 256, 384, 512, 768, 1024, 1536, 2048]:
>     x = torch.randn(num_tokens, hidden_size, dtype=torch.bfloat16, device=device)
>     weight = torch.randn(hidden_size, dtype=torch.bfloat16, device=device)
>     bias = torch.randn(hidden_size, dtype=torch.bfloat16, device=device)
>     eps = 1e-5
>     out = torch.empty_like(x)
> 
>     def native():
>         return F.layer_norm(x, (hidden_size,), weight, bias, eps)
> 
>     def sycl():
>         torch.ops._C.layer_norm(out, x, weight, bias, eps)
>         return out
> 
>     n = bench(native)
>     s = bench(sycl)
>     print(f"{num_tokens:<10}{n:<14.2f}{s:<14.2f}{n/s:<10.2f}x")
> ```
> 
> Notes for anyone reproducing:
> 
> * Wrote this standalone script during kernel development to check the kernel's raw compute performance in isolation, decoupled from the `CustomOp` dispatch/allocation overhead in `tests/ops/layernorm_op.py` — that machinery wasn't the point of the check, the kernel's own numbers were.
> * These are single-run means (300 iters, 30 warmup), no repeated-run median/spread — the `tokens=1` row is noisiest since actual compute time there is tiny enough that host-scheduling jitter dominates. Reran it independently and got the same numbers within ~1% at 128+ tokens; `tokens=1` varied more run to run.
> * This measures the raw op call — no `CustomOp` dispatch overhead, and the output buffer is preallocated once and reused across iterations. If you benchmark through the model-integration path (`XPULayerNorm` / `nn.LayerNorm` swap), expect a smaller speedup — the dispatch layer and a per-call output allocation add fixed overhead that's most visible at small `hidden_size`/token counts.
> * `torch.nn.functional.layer_norm` is the correct native comparison point here — it's the real fallback this kernel replaces, not the test suite's unfused fp32 reference implementation.

Please add benchmark file to here https://github.com/vllm-project/vllm-xpu-kernels/tree/main/benchmark.

**Comment by mganczarenko** [comment]
@mayuyuace, done

CC: @jikunshang

**Review by mayuyuace — APPROVED** [review]
